### ISSUE TITLE
JavaScript's TypeError is not a Ruby's TypeError (nor a Ruby's StandardError) ?

### ISSUE THREAD

**Issue by Mogztter** [opal/opal#1675]
Here's a simple way to reproduce this behavior:

.test.rb
```ruby
begin
  `null.oups`
rescue StandardError
  p 'rescue'
end
```

```
bundle exec opal -- test.rb
```

```
/tmp/opal-nodejs-runner-20170604-17922-10yevpr:21161
    } else { throw $err; }
             ^

TypeError: Cannot read property 'oups' of null
    at /tmp/opal-nodejs-runner-20170604-17922-10yevpr:21155:16
    at Object.<anonymous> (/tmp/opal-nodejs-runner-20170604-17922-10yevpr:21163:3)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:425:7)
    at startup (bootstrap_node.js:146:9)
```
The definition of JavaScript's TypeError and Ruby's TypeError are quite similar:

> The TypeError object represents an error when a value is not of the expected type.

https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/TypeError

> Raised when encountering an object that is not of the expected type.

https://ruby-doc.org/core-2.2.0/TypeError.html

As a workaround it's possible to `rescue Exception` but I think JavaScript's TypeError should be "map" as a Ruby's TypeError.

**Comment by noah538** [comment]
Is there any workaround for this TypeError? Tkx.

1. On my Win7 => working fine!
Asciidoctor.js Live Preview 1.5.5.100
asciidoc js preview
chrome: 59.0.3071.86 (standard version) (x64)

2. On Win10  =>  PREVIEW Error: TypeError : Cannot read property 'join' of null
Asciidoctor.js Live Preview 1.5.5.100
chrome: 59.0.3071.86 (standard version) (x64)


---------------------------------------------------------------------------
the console backtrace:

chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/renderer.js:64 TypeError: Cannot read property 'join' of null
    at Date.ːzone [as $zone] (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:17868)
    at chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:18056
    at RegExp.[Symbol.replace] (<anonymous>)
    at String.replace (<anonymous>)
    at Date.ːstrftime [as $strftime] (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:17919)
    at $Document.ːinitialize [as $initialize] (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:27367)
    at singleton_class_alloc.TMP_4 [as $new] (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:3146)
    at module_constructor.ːload [as $load] (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:33226)
    at module_constructor.Opal.Asciidoctor.$$class.$$proto.load (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/vendor/asciidoctor.js:33467)
    at updateBody (chrome-extension://iaalpfgpbocpdfblpnhhgllgbdbchmia/js/renderer.js:74)

**Comment by iliabylich** [comment]
@Mogztter I agree, these errors look similar, but I'm not sure that we should merge them.

Let's say we merge them and by writing `rescue TypeError` you catch both a JavaScript's and Ruby's TypeError.

How are you going to write a code that catches **only** Ruby's TypeError?
For example, `Struct` [relies](https://github.com/opal/opal/blob/master/opal/corelib/struct.rb#L10) on catching Ruby's TypeError. Having a broken JS code inside `begin .. rescue TypeError` code (like `null.m`) may hide a real error.

To catch JS errors there's an alternative - [`rescue JS::Error`](https://github.com/opal/opal/blob/master/opal/corelib/runtime.js#L1431) (added in https://github.com/opal/opal/pull/1561/files, you can check specs to get how it works) or writing something like
``` ruby
def method_that_catches_ruby_and_js_errors
  %x{
    try {
      #{
        begin
          throwing_code
        rescue RubyError1, RubyError2
          handle_ruby_error
        end
      }
    } catch (e) {
      handle_js_error
    }
  }
end
```

@elia @meh what do you think?

**Comment by elia** [comment]
@Mogztter @iliabylich completely agree.

Btw both 0.10 and 0.11 are able to rescue JS errors:

```
⤑ opal -ve"begin;  `null.oups`;rescue JS::Error;  p 'rescue';end"
Opal v0.11.0.rc1
"rescue"
```

```
⤑ opal -ve"begin;  `null.oups`;rescue Exception;  p 'rescue';end"
Opal v0.10.4
"rescue"
```

I'll close this for now, will reopen on request if something else comes up.

**Comment by Mogztter** [comment]
It's fine, we need to fix the root cause, plus I didn't think about catching Ruby's TypeError or JavaScript's TypeError (good point @iliabylich)

**Comment by elia** [comment]
@Mogztter the root cause should be fixed by b7701da and caaa3a2 😉

**Comment by Mogztter** [comment]
@elias Thank you very much :+1:
